RWAHS-596 Fix search by access point

leftclickben opened the pull request:
Fix issue with search by access points.
- Previously, an access point used in a search query would generate
  an array of `terms` and an array of `signs` that go with those `terms`,
  with the logical operator based on the definition of the access point 
  (`OR` by default).
- However this is incorrect, because the logical operator inside the
  access point was "leaking" out into the remainder of the query.  For
  example, if `ap` is an access point that consists attributes `x` and,
  `y`, and `z` is another attribute, then a search for 
  `ap:"foo" AND z:"bar"` would work as expected, but
  `z:"bar" AND ap:"foo"` (which should be the same query) would be
  converted into `z:"bar" OR x:"foo" OR y:"foo"`.
- In the above example, the `OR` is coming from the access point
  definition (or the default).  Also note the lack of parentheses.
- This changes the array to a `MultiTerm` Lucene object, which makes
  `z:"bar" AND ap:"foo"` convert to `z:"bar" AND (x:"foo" OR y:"foo")`,
  which is the intended behaviour.  Note the added parentheses.

Also fix a coding / typo issue in `LuceneSyntaxParserContext`, where a non-exception class was being `throw`n.

Commit:
Fix incorrect exception class reference.

Commit:
Fix issue with search by access points.

+ Previously, an access point used in a search query would generate
  an array of `terms` and an array of `signs` that go with those `terms`,
  with the logical operator based on the definition of the access point
  (`OR` by default).
+ However this is incorrect, because the logical operator inside the
  access point was "leaking" out into the remainder of the query.  For
  example, if `ap` is an access point that consists attributes `x` and,
  `y`, and `z` is another attribute, then a search for
  `ap:"foo" AND z:"bar"` would work as expected, but
  `z:"bar" AND ap:"foo"` (which should be the same query) would be
  converted into `z:"bar" OR x:"foo" OR y:"foo"`.
+ In the above example, the `OR` is coming from the access point
  definition (or the default).  Also note the lack of parentheses.
+ This changes the array to a `MultiTerm` Lucene object, which makes
  `z:"bar" AND ap:"foo"` convert to `z:"bar" AND (x:"foo" OR y:"foo")`,
  which is the intended behaviour.  Note the added parentheses.

kehh commented on app/lib/core/Search/SearchEngine.php:
```diff
@@ -548,16 +531,15 @@ private function _rewriteTerm($po_term, $pb_sign) {
 		}
 		
-		return array('terms' => array($po_term->getTerm()), 'signs' => array($pb_sign), 'options' => array());
+		return new Zend_Search_Lucene_Search_Query_MultiTerm(array($po_term->getTerm()), array($pb_sign));
```
Does this always have to be a `MultiTerm`? I'm just wondering whether the phpdoc should specify the parent class / interface?

leftclickben replied:
Updated as per your comment, which was related to the test failure.

kehh reviewed (commented):
Just the one question.

Commit:
Only return `MultiTerm` when necessary.

+ Previous change returned a `MultiTerm` from `_rewriteTerm()`, for
  every term, regardless of the number of actual terms.
+ This caused an extra level of nesting, which was causing tests to fail
  (and presumably other issues that hadn't been noticed through the UI).
+ This change makes it only return a `MultiTerm` when there are actually
  multiple terms.
+ Update the phpdoc to specify the `return` type as the abstract base
  class shared by `Term` and `MultiTerm`.
